File: src/javascript/vulnerabilities/sensitiveData.js
const { makeApiRequest } = require('../core/utils');

async function check(targetUrl, config) {
    const endpointsToCheck = config?.sensitive_data?.endpoints || [
        "/users",
        "/profile",
        "/account"
    ];
    
    const results = [];
    for (const endpoint of endpointsToCheck) {
        const response = await makeApiRequest(
            `${targetUrl}${endpoint}`,
            'GET'
        );
        
        if (response.error) {
            results.push({
                endpoint,
                error: response.error
            });
        } else {
            const sensitiveFields = [];
            const content = JSON.stringify(response.content || '');
            
            if (content.includes('password')) sensitiveFields.push('password');
            if (content.includes('credit_card')) sensitiveFields.push('credit_card');
            if (content.includes('ssn')) sensitiveFields.push('ssn');
            
            results.push({
                endpoint,
                statusCode: response.statusCode,
                sensitiveDataExposed: sensitiveFields.length > 0,
                fieldsFound: sensitiveFields
            });
        }
    }
    
    return {
        name: "Sensitive Data Exposure",
        description: "Checks for exposed sensitive data",
        results,
        severity: "High"
    };
}

module.exports = { check };

File: src/python/vulnerabilities/sensitive_data.py
from __future__ import annotations

import re
from typing import Any

from core.utils import build_url, get_request_options, make_api_request


def check(target_url: str, config: dict[str, Any]) -> dict[str, Any]:
    data_config = config.get("sensitive_data", {})
    endpoints_to_check = data_config.get("endpoints", [])
    patterns = data_config.get("patterns", {})
    compiled_patterns = {
        name: re.compile(pattern, re.IGNORECASE)
        for name, pattern in patterns.items()
    }
    request_options = get_request_options(config)

    results = []
    for endpoint in endpoints_to_check:
        response = make_api_request(
            build_url(target_url, endpoint),
            method="GET",
            **request_options,
        )

        if "error" in response:
            results.append({"endpoint": endpoint, "error": response["error"], "vulnerable": False})
            continue

        content = response.get("content", "") or ""
        fields_found = [name for name, pattern in compiled_patterns.items() if pattern.search(content)]
        exposed = response.get("status_code") in {200, 201, 202} and bool(fields_found)
        results.append({
            "endpoint": endpoint,
            "status_code": response.get("status_code"),
            "elapsed_ms": response.get("elapsed_ms"),
            "sensitive_data_exposed": exposed,
            "fields_found": fields_found,
            "vulnerable": exposed,
        })

    return {
        "name": "Sensitive Data Exposure",
        "description": "Checks configured endpoints for common sensitive fields and secrets in unauthenticated responses.",
        "severity": "High",
        "finding": any(result.get("vulnerable") for result in results),
        "results": results,
    }
